train.py

```diff
@@ -0,0 +1,249 @@
+    criterion = LabelSmoothing(size=len(vocab_tgt), padding_idx=pad_idx, smoothing=0.1)
+    if device != "cpu":
+        criterion.cuda(device)
+    # Create dataloaders
+    train_dataloader, valid_dataloader = create_dataloaders(
+        vocab_src,
+        vocab_tgt,
+        spacy_de,
+        spacy_en,
+        device=device,
+        slice=cfg.data_slice,
+        batch_size=cfg.batch_size,
+        max_padding=cfg.max_padding,
+    )
+
+    # Define optimizer and learning rate scheduler
+    optimizer = torch.optim.Adam(model.parameters(), lr=cfg.base_lr, betas=(0.9, 0.98), eps=1e-9)
+    lr_scheduler = LambdaLR(
+        optimizer=optimizer,
+        lr_lambda=lambda step: rate(step, cfg.model.d_model, factor=1, warmup=cfg.warmup),
+    )
+
+    train_state = TrainState()
+
+    for epoch in range(cfg.num_epochs):
+        model.train()
+        print(f"Epoch {epoch}", flush=True)
+
+        # Train model for one epoch
+        t_loss, train_state = train_epoch(
+            train_dataloader,
+            model,
+            criterion,
+            optimizer,
+            lr_scheduler,
+            pad_idx=pad_idx,
+            train_state=train_state,
+            accum_iter=cfg["accum_iter"],
+        )
+
+        # Save checkpoint after each epoch
+        file_path = f"models/{cfg.file_prefix}-{run_id}-{epoch}.pt"
+        torch.save(
+            {
+                "model_state_dict": module.state_dict(),
+                "cfg": OmegaConf.to_container(cfg),
+            },
+            file_path,
+        )
+        torch.cuda.empty_cache()
+        torch.cuda.empty_cache()
+        torch.cuda.empty_cache()
+
+        print(f"Epoch {epoch} Validation", flush=True)
+        model.eval()
+        # Evaluate the model on validation set
+        sloss = val_epoch(
+            train_dataloader,
+            model,
+            criterion,
+        )
+        # Log validation and training losses
+        print(sloss)
+        wandb.log({"val/loss": sloss, "train/loss": t_loss})
+        torch.cuda.empty_cache()
+        torch.cuda.empty_cache()
+        torch.cuda.empty_cache()
+
+    return model, run_id
+
+
+def train_epoch(
+    data_iter: Iterable,
+    model: nn.Module,
+    criterion: Callable,
+    optimizer: torch.optim.Optimizer,
+    scheduler: LambdaLR,
+    train_state: TrainState,
+    pad_idx=2,
+    accum_iter=1,
+) -> tuple[float, TrainState]:
+    start = time.time()
+    total_tokens = 0
+    total_loss = 0
+    tokens = 0
+    n_accum = 0
+    i = 0  # iteration counter
+
+    # create progress bar
+    pbar = tqdm(data_iter)
+    for b in pbar:  # for batch in dataloader
+        batch = Batch(b[0], b[1], pad_idx)
+        encode_decode = model.forward(batch.src, batch.tgt, batch.src_mask, batch.tgt_mask)
+        out = model.generator(encode_decode)
+        loss = criterion(einops.rearrange(out, "b n d -> (b n) d"), einops.rearrange(batch.tgt_y, "b n -> (b n)")) / batch.ntokens
```

Comment:
```suggestion
        out = einops.rearrange(out, "b n d -> (b n) d")
        target = einops.rearrange(batch.tgt_y, "b n -> (b n)")
        loss = criterion(out, target) / batch.ntokens
```

Please avoid nested function calls